apps/webapp/src/components/staking/use-validator-infos.ts

```diff
@@ -0,0 +1,77 @@
+import {
+  getValidatorInfo,
+  getVotingPowerFromValidatorInfo,
+  joinLoHiAmount,
+} from '@penumbra-zone/types';
+import { ValidatorInfo } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/component/stake/v1/stake_pb';
+import { useCollectedStream } from '../../fetchers/stream';
+import { getValidatorInfos } from '../../fetchers/staking';
+import { useMemo, useRef } from 'react';
+
+const byVotingPower = (validatorInfoA: ValidatorInfo, validatorInfoB: ValidatorInfo) =>
+  Number(
+    joinLoHiAmount(getVotingPowerFromValidatorInfo(validatorInfoB)) -
+      joinLoHiAmount(getVotingPowerFromValidatorInfo(validatorInfoA)),
+  );
+
+const toTotalVotingPower = (prev: number, curr: ValidatorInfo) =>
+  prev + Number(joinLoHiAmount(getVotingPowerFromValidatorInfo(curr)));
+
+const getFormattedVotingPower = (validatorInfo: ValidatorInfo, totalVotingPower: number) =>
+  Math.round(
+    (Number(joinLoHiAmount(getVotingPowerFromValidatorInfo(validatorInfo))) / totalVotingPower) *
+      100,
+  );
+
+export const useValidatorInfos = (): {
+  validatorInfos: ValidatorInfo[];
+
+  /** Each validator's voting power, expressed as an integer between 0-100. */
+  votingPowerByValidatorInfo: Map<ValidatorInfo, number>;
+
+  loading: boolean;
+  error: unknown;
+} => {
+  /**
+   * Use a ref so that it doesn't constantly re-fetch when the component
+   * re-renders.
+   *
+   * @todo: How to trigger a refresh?
+   */
+  const query = useRef(getValidatorInfos());
```

Comment:
I have psychological trauma about hooks calling hooks calling hooks (all those unmaintainable frontends I left behind 😅 ).

What do you think about moving all this to a react-router loader? (using async/await instead of hooks)

However, if we anticipate potentially slow or very long response, then it may not be best.

Reply:
Yeah, that was my original approach: I called `getValidatorInfos()` in the loader and then waited for the stream to finish. That took quite a while, though: testnet is a bit slow right now, and there are like 80 validators (300 if you include inactive ones, which I imagine we'll eventually do). So I figured we should stick with streaming the results.

Reply:
I don't think we need to support inactive validators in the list, only supporting members of the consensus set (and any validators the user has delegation tokens for) is fine.